File: Regex to NFA (Python code with Visualizations)/nfa.py
from collections import defaultdict
from queue import Queue

class NFA():
 
	characters = set()
	set_of_states = set()
	accepting_states = set()
	current_state = 0
	state_num = 0
	transition_function = defaultdict(set)
	RE = "" # regular expression
 
	# constructor
	def __init__(self, RE):
		self.RE = RE
	
	# build the alphabet from the characters
	def build_alphabet(self):
		for ch in range(0, len(self.RE)):
			if self.RE[ch] != '+' and self.RE[ch] != '*' and self.RE[ch] != '(' and self.RE[ch] != ')' and self.RE[ch] != ' ':
				self.characters.add(self.RE[ch])
		self.characters.add('E') # for the epsilon transitions
	
	# utility function for the recursion while building the NFA
	def generate_NFA(self, sub_initial_state, i):
		close_state = 0
		while i < len(self.RE):
			if self.RE[i] in self.characters and i < len(self.RE):
				if self.RE[i+1] != '*':
					self.transition_function[(self.current_state, self.RE[i])].add(self.state_num+1)
					self.current_state = self.state_num + 1
					self.state_num = self.state_num + 1
					self.set_of_states.add(self.current_state)
				else:
					self.transition_function[(self.current_state, self.RE[i])].add(self.state_num)
					i = i + 1
	 
			elif self.RE[i] in self.characters:
				if close_state != 0:
					self.transition_function[(self.current_state, self.RE[i])].add(self.state_num+1)
					self.current_state = self.state_num + 1
					self.state_num = self.state_num + 1
					self.set_of_states.add(self.current_state)
	 
			elif self.RE[i] == '+':
				if close_state == 0:
					self.current_state = self.state_num + 1
					self.state_num = self.state_num + 1
					self.set_of_states.add(self.current_state)
					close_state = self.current_state
					self.current_state = sub_initial_state
					self.transition_function[(self.state_num-1, 'E')].add(close_state)
				else:
					self.transition_function[(self.state_num, 'E')].add(close_state)
					self.current_state = sub_initial_state
	 
			elif self.RE[i] == '(':
				_sub_initial_state = self.current_state
				i = self.generate_NFA(_sub_initial_state, i+1)

			else:
				# self.RE[i] == ')'
				if i < len(self.RE)-1 and self.RE[i+1] == '*':
					if close_state == 0:
						self.transition_function[(self.state_num, 'E')].add(sub_initial_state)
					else:
						self.transition_function[(self.current_state, 'E')].add(close_state)
						self.transition_function[(close_state, 'E')].add(sub_initial_state)
					i = i + 1
					self.current_state = sub_initial_state
				else:
					if close_state != 0:
						self.transition_function[(self.current_state, 'E')].add(close_state)
						self.current_state = close_state
				return i
			i = i + 1
 
	def build_NFA(self):
		self.build_alphabet()
		self.set_of_states.add(0)
		initial_state = 0
		close_state = 0
  
		i = 0
		while i < len(self.RE):
			if self.RE[i] in self.characters and i < len(self.RE):
				if self.RE[i+1] != '*':
					self.transition_function[(self.current_state, self.RE[i])].add(self.state_num+1)
					self.current_state = self.state_num + 1
					self.state_num = self.state_num + 1
					self.set_of_states.add(self.current_state)
				else:
					self.transition_function[(self.current_state, self.RE[i])].add(self.state_num)
					i = i + 1
	 
			elif self.RE[i] in self.characters:
				if close_state != 0:
					self.transition_function[(self.current_state, self.RE[i])].add(self.state_num+1)
					self.current_state = self.state_num + 1
					self.state_num = self.state_num + 1
					self.set_of_states.add(self.current_state)
					
					if i == len(self.RE)-1:
						self.transition_function[(self.current_state, 'E')].add(close_state)
						self.current_state = close_state
				else:
					self.transition_function[(self.current_state, self.RE[i])].add(self.state_num+1)
					self.current_state = self.state_num + 1
					self.state_num = self.state_num + 1
					self.set_of_states.add(self.current_state)
	 
			elif self.RE[i] == '+':
				if close_state == 0:
					self.current_state = self.state_num + 1
					self.state_num = self.state_num + 1
					self.set_of_states.add(self.current_state)
					close_state = self.current_state
					self.current_state = initial_state
					self.transition_function[(self.state_num-1, 'E')].add(close_state)
					self.accepting_states.add(self.state_num)
				else:
					self.transition_function[(self.state_num, 'E')].add(close_state)
					self.current_state = initial_state
	 
			elif self.RE[i] == '(':
				sub_initial_state = self.current_state
				i = self.generate_NFA(sub_initial_state, i+1)
				if ((i == len(self.RE)-1 and self.RE[i] == '*') or (self.RE[i] == '*' and i < len(self.RE)-1 and self.RE[i+1] not in self.characters)):
					self.accepting_states.add(self.current_state)
				close_state = 0
			i = i + 1
	
		self.accepting_states.add(self.current_state)
  
	def is_valid_utility(self, ch, state):
		frontier = set()
		if (state, ch) in self.transition_function:
			next_states = self.transition_function[(state, ch)]
			frontier = frontier | next_states
   
		if (state, 'E') in self.transition_function:
			eclose = self.find_epsilon_closure(state)
			temp2 = set()
			if state not in eclose:
				temp2 = temp2 | eclose
			for st in temp2:
				frontier = frontier | self.transition_function[(st, ch)]
	
		return frontier

	def find_epsilon_closure(self, state):
		eclose = set()
		if (state, 'E') in self.transition_function:
			eclose = eclose | self.transition_function[(state, 'E')]
		for st in eclose:
			if (st, 'E') in self.transition_function:
				eclose = eclose | self.transition_function[(st, 'E')]
		return eclose

	def epsilon_transition_on_state(self, states):
		state = set()
		state.add(states)
		if (states, 'E') in self.transition_function:
			for s in self.transition_function[(states, 'E')]:
				state.add(s)
				temp = self.epsilon_transition_on_state(s)
				state.union(temp)
		return state
  
	def is_valid(self, w):
		for i in range(0, len(w)):
			if w[i] not in self.characters:
				return False
		i = 1
		final_states = self.is_valid_utility(w[0], 0)
		fs = final_states
		for st in fs:
			tempo = self.find_epsilon_closure(st)
			final_states = final_states | tempo

		while(len(final_states) > 0 and i < len(w)):
			temp_final = set()
			for state in final_states:
				temp = self.is_valid_utility(w[i], state)
				temp_final = temp_final | temp
			final_states = temp_final
			i = i + 1
   
		temporary = set()
		for state in final_states:
			temps = self.find_epsilon_closure(state)
			temporary = temporary | temps
		final_states = final_states | temporary
  
		for state in final_states:
			if state in self.accepting_states:
				return True
		return False

	# shortest string to any accepting state from a given start state
	def Dijkstra(self, start_state):
		strings = []
		for i in range(0, len(self.set_of_states)):
			strings.append("+")
		strings[start_state] = "" # empty string to itself
  
		Q = Queue()
		Q.put(start_state)
		while not Q.empty():
			state = Q.get()
			for ch in self.characters:
				if (state, ch) in self.transition_function:
					# relax for all out-neighbours
					out_neighbours = set()
					out_neighbours = out_neighbours | self.transition_function[(state, ch)]
					for v in out_neighbours:
						y = strings[state] + ch
						if ch == 'E': 
							y = strings[state]
						z = strings[v]
						if z == '+':
							strings[v] = y
							Q.put(v)
						else:
							if len(strings[v]) > len(y):
								strings[v] = y
	
		it = iter(self.accepting_states)
		first_element = next(it)
		shortest = strings[first_element]
		for accept in self.accepting_states:
			if len(strings[accept]) < len(shortest):
				shortest = strings[accept]
		return shortest

	def print_transitions(self):
		for key, value in self.transition_function.items():
			print(key, value)

	def print_accepting_states(self):
		print("Accepting states are: ", end="")
		print(self.accepting_states)

File: nfa.cpp
#include <iostream>
#include <set>
#include <map>
#include <utility>
#include <string>
#include <iterator>
#include <queue>
using namespace std;

class NFA{
    private:
        int currentState;
        // utility functions

        // Building Alphabets
        void buildAlphabet(string RE){
            RE = compactRegexGeneration(RE);
            for(int i = 0; i < RE.size(); i++){
                if(RE[i] != '+' && RE[i] != '*' && RE[i] != '(' && RE[i] != ')' && RE[i] != ' ') characters.insert(RE[i]);
                characters.insert('E');
            }
        }

        // Compact form of REGEX
        string compactRegexGeneration(string Regex){
        	string compact="";
	        for(int i=0;i<Regex.size();i++){
		        if(Regex[i] == ' ') continue;
    		    compact+=Regex[i];
	        }
	        return compact;
        }
    
        int generateNFA(int subInitialState, string RE, int i){
            int closeState = 0;
            for(;i < RE.size(); i++){
                if(characters.find(RE[i]) != characters.end() && i < RE.size()-1){
                    if(RE[i+1] != '*') {
                        transitionFunction[make_pair(currentState,RE[i])].insert(stateNum+1);
                        currentState=++stateNum;
                        setOfStates.insert(currentState);
                    }
                    else{
                        transitionFunction[make_pair(currentState,RE[i])].insert(stateNum);
                        i++;
                    }
                }
                else if(characters.find(RE[i]) != characters.end()) {
                    transitionFunction[make_pair(stateNum,RE[i])].insert(stateNum++);
                    currentState = stateNum;
                    setOfStates.insert(currentState);
                }
                else if(RE[i] == '+'){

                    if(closeState == 0){
                        currentState = ++stateNum;
                        setOfStates.insert(currentState);
                        closeState = currentState;
                        currentState = subInitialState;
                        transitionFunction[make_pair(stateNum-1, 'E')].insert(closeState);
                    } else{
                        transitionFunction[make_pair(stateNum, 'E')].insert(closeState);
                        currentState = subInitialState;
                    }
                }
                else if(RE[i] == '('){
                    int _subInitialState = currentState;
                    i = generateNFA(_subInitialState, RE, i+1);
                }
                else if(RE[i] == ' ') continue;
                else{ // RE[i] = ')'
                    if(i < RE.size()-1 && RE[i+1] == '*'){
                        if(closeState == 0) {
                            transitionFunction[make_pair(stateNum,'E')].insert(subInitialState);
                        }
                        else{
                            transitionFunction[make_pair(currentState, 'E')].insert(closeState);
                            transitionFunction[make_pair(closeState,'E')].insert(subInitialState);
                        }
                        i += 1;
                        currentState = subInitialState;
                    }
                    else{
                        if(closeState != 0) {
                            transitionFunction[make_pair(currentState, 'E')].insert(closeState);
                            currentState = closeState;
                        }
                    }
                    return i;
                }
            }
        }

        set<int> isValidUtility(char w, int state){
            set<int> frontier;
            if(transitionFunction.find(make_pair(state, w)) != transitionFunction.end()){
                frontier.insert(transitionFunction[make_pair(state, w)].begin(), transitionFunction[make_pair(state, w)].end());
            }

            if(transitionFunction.find(make_pair(state, 'E')) != transitionFunction.end()){
                //set<int>::iterator itr = transitionFunction[make_pair(state, 'E')].begin();
                set<int> eclose = findEpsilonClosure(state);
                if(eclose.find(state) == eclose.end()){
                    frontier.insert(eclose.begin(), eclose.end());
                }

                /*while(itr!=transitionFunction[make_pair(state, 'E')].end()){
                    set<int> temp = isValidUtility(w,*itr);
                    frontier.insert(temp.begin(), temp.end());
                    itr++;
                }*/
            }
            cout << "cdvf";
            printSetOfStates(frontier);
            return frontier;
        }

        set<int> findEpsilonClosure(int state){
            set<int> eclose;
            set<int>::iterator itr;
            if(transitionFunction.find(make_pair(state, 'E')) != transitionFunction.end())
                eclose.insert(transitionFunction[make_pair(state, 'E')].begin(), transitionFunction[make_pair(state, 'E')].end());
            for(itr = eclose.begin(); itr != eclose.end(); itr++){
                if(transitionFunction.find(make_pair(*itr, 'E')) != transitionFunction.end())
                eclose.insert(transitionFunction[make_pair(*itr, 'E')].begin(), transitionFunction[make_pair(*itr, 'E')].end());
            }
            printSetOfStates(eclose);
            return eclose;
        }

        set<int> epsilonTransitionsOnState(int states){
            set<int> state;
            state.insert(states);
            if(transitionFunction.find(make_pair(states, 'E')) != transitionFunction.end()){
                set<int>::iterator itr = transitionFunction[make_pair(states, 'E')].begin();
                for(; itr != transitionFunction[make_pair(states, 'E')].end(); itr++){
                    state.insert(*itr);
                    set<int> temp = epsilonTransitionsOnState(*itr);
                    state.insert(temp.begin(), temp.end());
                }
            }
            return state;
        }

    public:
        // 5-tuple
        set<int> setOfStates;
        set<char> characters;
        set<int> acceptingStates;
        int startState;
        map<pair<int, char>, set<int> > transitionFunction;

        NFA(){

        }

        int stateNum;
        void buildNFA(string RE){
            buildAlphabet(RE);
            stateNum = 0;
            startState = 0; setOfStates.insert(0);
            currentState = 0;

            int initialState = 0; int closeState = 0;
            for(int i = 0; i < RE.size(); i++){
                if(characters.find(RE[i]) != characters.end() && i < RE.size()-1){
                    if(RE[i+1] != '*') {
                        transitionFunction[make_pair(currentState,RE[i])].insert(stateNum+1);
                        currentState=++stateNum;
                        setOfStates.insert(currentState);
                    }
                    else{
                        transitionFunction[make_pair(currentState,RE[i])].insert(stateNum);
                        i++;
                    }
                }
                else if(characters.find(RE[i]) != characters.end()) {
                    if(closeState != 0){
                        transitionFunction[make_pair(currentState,RE[i])].insert(stateNum+1);
                        currentState = ++stateNum;
                        setOfStates.insert(currentState);
                        if(i == RE.size()-1){
                            transitionFunction[make_pair(currentState, 'E')].insert(closeState);
                            currentState = closeState;
                        }
                    }
                    else{
                        transitionFunction[make_pair(currentState,RE[i])].insert(stateNum+1);
                        currentState = ++stateNum;
                        setOfStates.insert(currentState);
                    }
                }
                else if(RE[i] == '+'){

                    if(closeState == 0){
                        currentState = ++stateNum;
                        setOfStates.insert(currentState);
                        closeState = currentState;
                        currentState = initialState;
                        transitionFunction[make_pair(stateNum-1, 'E')].insert(closeState);
                        acceptingStates.insert(stateNum);
                    }
                    else{
                        transitionFunction[make_pair(stateNum, 'E')].insert(closeState);
                        currentState = initialState;
                    }
                }
                else if(RE[i] == '('){
                    int subInitialState = currentState;

                    i = generateNFA(subInitialState, RE, i+1);
                    if((RE[i] == '*' && i == RE.size()-1) || (RE[i] == '*' && i < RE.size()-1 && characters.find(RE[i+1]) == characters.end()))
                        acceptingStates.insert(currentState);
                    closeState = 0;
                }
            }
            acceptingStates.insert(currentState);
        }

        bool isValid(string w){
            for(int i = 0; i < w.size(); i++){
                if(characters.find(w[i]) == characters.end()) return false;
            }

            set<int> finalStates = isValidUtility(w[0], 0);
            int i = 1;
            while(!finalStates.empty() && i < w.size()){
                set<int> tempFinal;
                set<int>::iterator itr = finalStates.begin();
                while(itr != finalStates.end()){
                    set<int> temp = isValidUtility(w[i],*itr);
                    tempFinal.insert(temp.begin(), temp.end());
                    itr++;
                }
                finalStates.swap(tempFinal);
                i++;
            }

            set<int> temporary;
            set<int>::iterator finalCheck = finalStates.begin();
            for(;finalCheck != finalStates.end(); finalCheck++){
                set<int> temps = epsilonTransitionsOnState(*finalCheck);
                temporary.insert(temps.begin(), temps.end());
            }
            temporary.swap(finalStates);
            finalCheck = finalStates.begin();
            for(; finalCheck != finalStates.end(); finalCheck++){
                if(acceptingStates.find(*finalCheck) != acceptingStates.end()) return true;
            }
            return false;
        }

        string Dijkstra(int beginState){
            vector<string> strings;
            for(int i = 0; i < setOfStates.size(); i++) strings.push_back("+");
            strings[beginState] = "";
            queue<pair<int, string> > Q;
            Q.push(make_pair(beginState, strings[beginState]));
            while(!Q.empty()){
                pair<int, string> x = Q.front();
                Q.pop();

                set<char>::iterator itr;
                for(itr = characters.begin(); itr != characters.end(); itr++){
                    if(transitionFunction.find(make_pair(x.first, *itr)) != transitionFunction.end()){
                        // relax for all out-neighbors
                        set<int> outNeighbours = transitionFunction[make_pair(x.first, *itr)];
                        set<int>::iterator ii;
                        for(ii = outNeighbours.begin(); ii != outNeighbours.end(); ii++){
                            string y = strings[x.first] + (*itr);
                            if(*itr == 'E') y = strings[x.first];
                            string z = strings[*ii];
                            if(z == "+"){
                                strings[*ii] = y;
                                Q.push(make_pair(*ii, strings[*ii]));
                            }
                            else{
                                if(strings[*ii].size() > y.size()) strings[*ii] = y;
                            }
                        }
                    }
                }
            }
            set<int>::iterator it = acceptingStates.begin();
            string shortest = strings[*it];
            for(; it != acceptingStates.end(); it++){
                if(shortest.size() > strings[*it].length()) shortest = strings[*it];
            }
            return shortest;
        }

        void printTransitionFunction(){
            map<pair<int, char>, set<int> >::iterator itr;
            set<int>::iterator delta;
            for(itr = transitionFunction.begin(); itr != transitionFunction.end(); itr++){
                cout << "(" << itr->first.first << ", " << itr->first.second << "): ";
                for(delta = itr->second.begin(); delta != itr->second.end(); delta++){
                    cout << *delta << " ";
                }
                cout << endl;
            }
            cout << endl;
        }

        void printSetOfStates(set<int> s){
            set<int>::iterator it = s.begin();
            for(; it != s.end(); it++) cout << *it << " ";
            cout << endl;
        }
};

int main(){
    NFA N;
    //N.buildNFA("aaab+ba(ab)*");
    N.buildNFA("(a+(ab)*)*");
    N.printTransitionFunction();
    cout << "Accepting State(s): ";
    N.printSetOfStates(N.acceptingStates);
    cout << endl;

    string w;
    cout << "Enter the string: ";
    cin >> w;
    if(N.isValid(w)) cout << w << " is accepted by the NFA\n";
    else cout << w << " is not accepted by the NFA\n";

    string shortest = N.Dijkstra(0);
    shortest = shortest=="+"||shortest=="" ? "the empty string" : shortest;
    cout << "The shortest string accepted by the NFA/Regex is: " << shortest << endl;
}
